File: lib/leetcode/680_valid-palindrome-ii.js
/**
 * @param {string} s
 * @return {boolean}
 */
var validPalindrome = function(s) {
  let head = 0;
  let tail = s.length - 1;
  // 'c bbc c'
  while (head < tail) {
    if (s[head] != s[tail]) {
      return (
        isPalindrome2(s.substring(head + 1, tail + 1)) ||
        isPalindrome2(s.substring(head, tail))
      );
    } else {
      head++;
      tail--;
    }
  }

  return true;
};

function isPalindrome(str) {
  console.log(str);

  return (
    str ===
    str
      .split("")
      .reverse()
      .join("")
  );
}

function isPalindrome2(str) {
  let head = 0;
  let tail = str.length - 1;

  while (head < tail) {
    if (str[head] !== str[tail]) {
      return false;
    } else {
      head++;
      tail--;
    }
  }
  return true;
}

console.log(isPalindrome("abbbbbac"));

let result = validPalindrome("cbbcc");
console.log(result);

File: lib/leetcode/680_valid-palindrome-ii.rb
def valid_palindrome(s)
  l = 0
  r = s.length - 1
  @str = s
  while l < r
    return palindrome(l, r - 1) || palindrome(l + 1, r) if s[l] != s[r]

    l += 1
    r -= 1
  end
  true
end

def palindrome(l, r)
  while l < r
    return false if @str[l] != @str[r]

    l += 1
    r -= 1
  end
  true
end
